en/Building_a_Simple_Engine/GUI/05_vulkan_integration.adoc

```diff
@@ -0,0 +1,1001 @@
+    ImGuiIO& io = ImGui::GetIO();
+
+    // Call input callback if registered
+    if (inputCallback) {
+        inputCallback(io);
+    }
+
+    ImGui::NewFrame();
+}
+
+void ImGuiUtil::Render(vk::raii::CommandBuffer& commandBuffer) {
+    if (!initialized) return;
+
+    ImGui::Render();
+    renderer.render(ImGui::GetDrawData(), commandBuffer);
+}
+
+void ImGuiUtil::UpdateDisplaySize(int width, int height, float scaleX, float scaleY) {
+    if (!initialized) return;
+
+    ImGuiIO& io = ImGui::GetIO();
+    io.DisplaySize = ImVec2(static_cast<float>(width), static_cast<float>(height));
+    io.DisplayFramebufferScale = ImVec2(scaleX, scaleY);
+}
+
+bool ImGuiUtil::ProcessInputEvent(void* event) {
+    // Platform-specific event processing would go here
+    // This is a placeholder for the actual implementation
+    return false;
+}
+
+void ImGuiUtil::SetInputCallback(std::function<void(ImGuiIO&)> callback) {
+    inputCallback = callback;
+}
+
+void ImGuiUtil::createDescriptorPool() {
+    // Tune these to match your expected number of UI textures and buffers.
+    vk::DescriptorPoolSize poolSizes[] =
+    {
+        { vk::DescriptorType::eSampler, 8 },
+        { vk::DescriptorType::eCombinedImageSampler, 128 },
+        { vk::DescriptorType::eSampledImage, 128 },
+        { vk::DescriptorType::eStorageImage, 8 },
+        { vk::DescriptorType::eUniformTexelBuffer, 8 },
+        { vk::DescriptorType::eStorageTexelBuffer, 8 },
+        { vk::DescriptorType::eUniformBuffer, 32 },
+        { vk::DescriptorType::eStorageBuffer, 32 },
+        { vk::DescriptorType::eUniformBufferDynamic, 16 },
+        { vk::DescriptorType::eStorageBufferDynamic, 16 },
+        { vk::DescriptorType::eInputAttachment, 8 }
+    };
+
+    uint32_t maxSets = 0;
+    for (const auto& ps : poolSizes) maxSets += ps.descriptorCount;
+
+    vk::DescriptorPoolCreateInfo poolInfo{
+        .flags = vk::DescriptorPoolCreateFlagBits::eFreeDescriptorSet,
+        .maxSets = maxSets,
+        .poolSizeCount = static_cast<uint32_t>(std::size(poolSizes)),
+        .pPoolSizes = poolSizes
+    };
+
+    descriptorPool = vk::raii::DescriptorPool(*device, poolInfo);
+}
+
+void ImGuiUtil::uploadFonts() {
+    vk::raii::CommandBuffer commandBuffer = beginSingleTimeCommands();
+    renderer.uploadFonts(commandBuffer);
+    endSingleTimeCommands(commandBuffer);
+}
+
+vk::raii::CommandBuffer ImGuiUtil::beginSingleTimeCommands() {
+    vk::CommandBufferAllocateInfo allocInfo{
+        .commandPool = *commandPool,
+        .level = vk::CommandBufferLevel::ePrimary,
+        .commandBufferCount = 1
+    };
+
+    vk::raii::CommandBuffer commandBuffer = vk::raii::CommandBuffers(*device, allocInfo).front();
+
+    vk::CommandBufferBeginInfo beginInfo{
+        .flags = vk::CommandBufferUsageFlagBits::eOneTimeSubmit
+    };
+
+    commandBuffer.begin(beginInfo);
+
+    return commandBuffer;
+}
+
+void ImGuiUtil::endSingleTimeCommands(vk::raii::CommandBuffer& commandBuffer) {
```

Comment:
Are `beginSingleTimeCommands` and `endSingleTimeCommands` really specific ImGuiUtil? Or could they be just some free functions?